core/plugin_impl_base.h

```diff
@@ -56,7 +59,7 @@ class PluginImplBase
     const PluginImplBase &operator=(const PluginImplBase &) = delete;
 
 protected:
-    Device &_parent;
+    std::shared_ptr<MAVLinkSystem> _parent;
```

Comment:
Why didn't you leave this a reference?

Reply:
class `System` creates a `shared_ptr` to `MAVLinkSystem` during construction to finally share them to various plugins. I made it so, because each plugin holds a shared instance of the system. It made sense to me. Do you see any issue ?

Reply:
I think a reference was fine, and generally you wanted to favor references against pointers, right?

Reply:
Yeah, I favor reference over a normal (raw) pointers. I can change to references as well.

Reply:
I guess the question of smart pointer vs reference is about ownership: say I have an object `A` that has an object `o`. If `o` only lives when `A` lives, then you can pass a reference to other objects using `o`. You want to have a pointer when ownership can change, i.e. when `o` may live longer than `A`. And there you have the question between `unique_ptr` and `shared_ptr`.

I am not completely sure of the design here, but it seems like a reference is what you want, indeed.

Reply:
I create `MAVLinkSystem` instance in [System constructor](https://github.com/dronecore/DroneCore/pull/310/files#diff-de9116ca3f3e99f357eb697180be37f6R22). I've to use pointer here. So, as it is going to be shared by all the plugins, I made it as `shared_ptr`. Now, while giving access to plugins, it is better to give them as `shared_ptr` itself, rather than dereference and give reference. What do you think ?

Reply:
@shakthi-prashanth-m why do you have to create `MAVLinkSystem` on the heap in the system constructor?

Reply:
So that `System` can hide `MAVLinkSystem` class. Now we have pointer to it.